common/sampling.cpp

```diff
@@ -41,7 +42,7 @@ struct llama_sampling_context * llama_sampling_init(const struct llama_sampling_
 
     result->n_valid = 0;
 
-    llama_sampling_set_rng_seed(result, params.seed);
+    llama_sampling_set_rng_seed(result->smpl, params.seed);
 
     return result;
 }
```

Comment:
Should `llama_sampling_free` be adjusted to also attempt to free `ctx->smpl`, similar to how it does for the grammar?

```
void llama_sampling_free(struct llama_sampling_context * ctx) {
    if (ctx->grammar != NULL) {
        llama_grammar_free(ctx->grammar);
    }
    if (ctx->smpl != NULL) {
        llama_sampling_free(ctx->smpl);
    }
    delete ctx;
}
```

It's a bit odd to me that the same function name (`llama_sampling_free`) is being used to free both the `llama_sampling_context` as well as `llama_sampling` objects. I still haven't fully wrapped my head around what a `llama_sampling` object _is_, so that might be part of my confusion as well.

So far this PR is looking good though! This is a really nice cleanup!

Reply:
Thanks for looking and suggesting improvements! The reason to not free `ctx->smpl` is because during init, we might pass a `llama_sampling` that we created separately and that `llama_sampling_context` will not "own". Similar change will be done for `ctx->grammar`. Will consider adding an init option where the `llama_sampling_context` creates it's own `smpl` and `grammar` and takes care to free them automatically for convenience.

Currently `llama_sampling_context` is badly named - it's becoming more of a wrapper of `llama_sampling` + `llama_grammar` + some stuff that I want to try to move into `llama_sampling`. I will probably change it's name soon since it's confusing at the moment and definitely should avoid the current function overloads